### DISCUSSION TITLE
Embedded emulator to run games & menus in the browser

### DISCUSSION THREAD

**Discussion by nguillaumin** [atari-legend/atari-legend#235] [Ideas]
Hi all,

I made a little proof of concept of embedding an Atari emulator in the site to run games & menus directly in the browser:

https://github.com/user-attachments/assets/092ec625-c5ef-45e8-8025-4fc64549f4d8

It works by compiling native programs written in C or C++ into HTML/JS (WebAssembly) via [Emscripten](https://emscripten.org/).

While it's pretty cool, unfortunately I don't see how to convert this proof of concept into something concrete. There are many challenges in finding the right emulator that will work for all use cases:
- In my test I used [PCEjs](https://jamesfriend.com.au/pce-js/atari-st/). It's based on [PCE](http://www.hampa.ch/pce/pce-atarist.html). It was the simplest to embed as they provide NPM packages (although I still struggled quite a bit!). But there are 2 major issues: 1/ it does not support sound and 2/ it does not support joysticks. 1/ would be bearable, but without 2/ no games can be played.
- There's a version of [Hatari](http://absencehq.de/union/) from the [forum](https://www.atari-forum.com/viewtopic.php?t=27255). It's also more or less a proof of concept. It's an old version from 10 years ago, and it doesn't work in Chrome (only Firefox). Part of the code was contributed back to Hatari though and it's supposed to be compilable via Emscripten, but it's not working for me I get an error towards the end that I don't know how to fix [1]. Even if I fix it, not sure how much works remain to make it useable in a browser.
- There's also MAME/MESS which has an Atari ST driver. I was able to compile it with Emscripten [after some work](https://github.com/mamedev/mame/issues/12669#issuecomment-2346742544), but it would take more work to make it work because the web compiled version cannot access files easily (e.g. TOS rom, IKDB rom, and floppy image), it needs a fake in-browser file system. I think that should be possible by taking some inspiration from [Emularity](https://github.com/db48x/emularity). But that's some amount of work, and we would need to be sure that this emulator works for our purposes (I don't know how accurate it is and if it generally works or has compatibility issues). If somebody has experience with it it would be great to share.

Overall trying to build and run things with Emscripten is a bit above my knowledge :smiley: . In any case with the web eating the world I think we will see more and more web-based emulator, and when a good Atari ST one emerges we will be able to embed it. Maybe all we have to do is wait a bit more...

---

[1] Hatari Emscripten build error:

```
[100%] Linking C executable hatari.html
Nothing to do!
emcc: error: '/usr/lib/emscripten/tools/file_packager hatari.data --from-emcc --preload /home/nicolas/Downloads/hatari/files/@/share/hatari/' failed (returned 1)
make[3]: *** [src/CMakeFiles/hatari.dir/build.make:1227: src/hatari.html] Error 1
make[2]: *** [CMakeFiles/Makefile2:394: src/CMakeFiles/hatari.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:401: src/CMakeFiles/hatari.dir/rule] Error 2
make: *** [Makefile:218: hatari] Error 2
emmake: error: 'make hatari -j 6' failed (returned 2)
```

**Comment by photonstorm** [comment]
The guys over at dbug got Hatari working well. Maybe drop them a line?

See here: https://d-bug.me/atariaviary/hatari.html?id=1

**Reply by stgraveyard** [reply]
Oh wow, this is working really well for me. Music, SFX and keyboard control. 
@nguillaumin should we get in touch with the DBUG guys?

**Reply by nguillaumin** [reply]
Hey, thanks for this, I didn't know it! I will get in touch with them.

**Reply by stgraveyard** [reply]
@MugUK do you have good relations with the DBUG guys? I think so, right? Can you maybe help out?

**Comment by stgraveyard** [comment]
So as @MugUK said, this is the true source link : [http://absencehq.de/atariaviary/](http://absencehq.de/atariaviary/?fbclid=IwZXh0bgNhZW0CMTAAAR0TRG3gX4gpvUbBkyjXGtC15N1dB36PNiGix03HJ1TmbnfJwh-0IF7ga3I_aem_S9vnd1UnZ2EUgaB_MZeZRw)

I think we can contact them ...

**Comment by ggnkua** [comment]
Pinging @tin-nl (ignoring the saltiness of the thread!)

**Comment by nguillaumin** [comment]
Thanks for the input guys. I wasn't aware of a "war" :wink: . I sent an email as well.

No need to contact Hatari I think. It can be compiled to a web version. The challenge is all the hard work "on the edge" to make it work that these guys already did and would be nice to reuse...

**Comment by stgraveyard** [comment]
![SC1435 green](https://github.com/user-attachments/assets/d2579001-b463-4a42-a08f-7a95deea339f)
![SC1435](https://github.com/user-attachments/assets/b9d50cda-4c9f-4abe-a8d9-55fb119c4ed8)
![SM124 no bg no shadow green](https://github.com/user-attachments/assets/be1e585c-66af-4722-bd54-e70bfdc057ef)
![SM124 no bg no shadow](https://github.com/user-attachments/assets/c3778dc0-ae73-4013-8bcf-6a41ff99745b)

Nico, some stuff to test with perhaps? If you need others, just let me know.

**Comment by stgraveyard** [comment]
@nguillaumin look at this : https://kaiec.github.io/EstyJS/?fbclid=IwY2xjawG1BeNleHRuA2FlbQIxMAABHaR8OgnHENfAdNAMSf8_1turvQf5bnhI_dBLRxtekI2jJqEWeTAsdArQ8Q_aem_PUAk9l1a2GRtK5YE4ESmew

New version!

**Comment by nguillaumin** [comment]
I could resume work on this with AI help. I deployed a version on the [dev site](https://dev.atarilegend.com/) that has the emulator working (from the AtariAviary site) . It's working both for games and menus. Whenever something is "playable" (i.e. a dump is available) there's a play icon in the screenshot that sends to the emulator.

<img width="523" height="617" alt="image" src="https://github.com/user-attachments/assets/6474db5c-e354-4dee-a64a-e731f4ed7182" />

<img width="962" height="600" alt="image" src="https://github.com/user-attachments/assets/51154d92-c95a-4e9a-bade-af9e4013317e" />


<img width="945" height="563" alt="image" src="https://github.com/user-attachments/assets/3c8db2f3-aa97-4c44-a6c7-7fa27fbaa2ab" />

The emulator page looks like this:

<img width="1866" height="1460" alt="image" src="https://github.com/user-attachments/assets/ec4bbb4f-f704-4046-b156-c1afb454c42a" />

Generally it works well, but there are still things to polish especially around compatibility, i.e. picking the right TOS version or memory size depending on the compatibility information in the AL database (we don't have that info for menus though, only for game releases, so not sure how to handle this yet - maybe let the user switch TOS versions?).

Have a look, let me know what you think.

FYI @stgraveyard
